smarts/core/sensors.py

```diff
@@ -588,68 +579,39 @@ def steps_completed(self):
 
 
 class CameraSensor(Sensor):
-    def __init__(self, vehicle, scene_np: NodePath, showbase: ShowBase):
-        self._log = logging.getLogger(self.__class__.__name__)
-        self._vehicle = vehicle
-        self._scene_np = scene_np
-        self._showbase = showbase
-
-    def _build_offscreen_camera(
+    def __init__(
         self,
+        vehicle,
+        renderer: Renderer,
         name: str,
         mask: int,
         width: int,
         height: int,
         resolution: float,
     ):
-        # setup buffer
-        win_props = WindowProperties.size(width, height)
-        fb_props = FrameBufferProperties()
-        fb_props.setRgbColor(True)
-        fb_props.setRgbaBits(8, 8, 8, 1)
-        # XXX: Though we don't need the depth buffer returned, setting this to 0
-        #      causes undefined behaviour where the ordering of meshes is random.
-        fb_props.setDepthBits(8)
-
-        buffer = self._showbase.win.engine.makeOutput(
-            self._showbase.pipe,
-            "{}-buffer".format(name),
-            -100,
-            fb_props,
-            win_props,
-            GraphicsPipe.BFRefuseWindow,
-            self._showbase.win.getGsg(),
-            self._showbase.win,
-        )
-        buffer.setClearColor((0, 0, 0, 0))  # Set background color to black
-
-        # setup texture
-        tex = Texture()
-        region = buffer.getDisplayRegion(0)
-        region.window.addRenderTexture(
-            tex, GraphicsOutput.RTM_copy_ram, GraphicsOutput.RTP_color
-        )
-
-        # setup camera
-        lens = OrthographicLens()
-        lens.setFilmSize(width * resolution, height * resolution)
-
-        camera_np = self._showbase.makeCamera(
-            buffer, camName=name, scene=self._scene_np, lens=lens
+        self._log = logging.getLogger(self.__class__.__name__)
+        self._vehicle = vehicle
+        self._camera = renderer.build_offscreen_camera(
+            name,
+            mask,
+            width,
+            height,
+            resolution,
         )
-        camera_np.reparentTo(self._scene_np)
 
-        # mask is set to make undesireable objects invisible to this camera
-        camera_np.node().setCameraMask(camera_np.node().getCameraMask() & mask)
+    def teardown(self):
+        self._camera.teardown()
 
-        return OffscreenCamera(camera_np, buffer, tex)
+    def step(self):
+        self._follow_vehicle()
 
-    def _follow_vehicle(self, camera_np):
+    def _follow_vehicle(self):
         center = self._vehicle.position
         largest_dim = max(self._vehicle._chassis.dimensions.as_lwh)
+        camera_np = self._camera.camera_np
         camera_np.setPos(center[0], center[1], 20 * largest_dim)
         camera_np.lookAt(*center)
```

Comment:
I think in this case it might be OK to continue to use the `NodePath` here directly. Cameras have a tendency to have privileged access to the renderer.